src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,234 @@
+
+#include <string.h>
+
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/RandUtils.h>
+
+using chip::Mdns::kMdnsTextMaxSize;
+using chip::Mdns::MdnsService;
+using chip::Mdns::TextEntry;
+
+namespace {
+
+void FreeTextEntries(MdnsService * service)
+{
+    if (service->mTextEntryies)
+    {
+        for (size_t i = 0; i < service->mTextEntrySize; i++)
+        {
+            TextEntry & entry = service->mTextEntryies[i];
+
+            if (entry.mKey)
+            {
+                chip::Platform::MemoryFree(const_cast<char *>(entry.mKey));
+            }
+            if (entry.mData)
+            {
+                chip::Platform::MemoryFree(const_cast<uint8_t *>(entry.mData));
+            }
+        }
+        chip::Platform::MemoryFree(service->mTextEntryies);
+    }
+}
+
+CHIP_ERROR CopyTextEntries(MdnsService * dest, const MdnsService & source)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (source.mTextEntrySize)
+    {
+        dest->mTextEntrySize = source.mTextEntrySize;
+        dest->mTextEntryies  = static_cast<TextEntry *>(chip::Platform::MemoryCalloc(source.mTextEntrySize, sizeof(TextEntry)));
+        VerifyOrExit(dest->mTextEntryies != nullptr, error = CHIP_ERROR_NO_MEMORY);
+
+        for (size_t i = 0; i < source.mTextEntrySize; i++)
+        {
+            TextEntry & entry = dest->mTextEntryies[i];
+
+            entry.mKey = static_cast<char *>(chip::Platform::MemoryAlloc(strnlen(source.mTextEntryies[i].mKey, kMdnsTextMaxSize)));
+            VerifyOrExit(entry.mKey != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            entry.mDataSize = source.mTextEntryies[i].mDataSize;
+            entry.mData     = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(entry.mDataSize));
+            VerifyOrExit(entry.mData != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::ServicePool()
+{
+    ClearNoFree();
+}
+
+CHIP_ERROR ServicePool::AddService(uint64_t nodeId, uint64_t fabricId, const MdnsService & service, bool deepCopy)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+    bool found       = false;
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    do
+    {
+        if (mEntries[i].mLazyDelete || mEntries[i].mNodeId == kUndefinedNodeId)
+        {
+            found = true;
+            break;
+        }
+
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    if (!found)
+    {
+        // Remove the entry that just matches the hash
+        i = hashValue;
+    }
+
+    if (mEntries[i].mLazyDelete)
+    {
+        mLazyDeleteCount--;
+        mEntries[i].mLazyDelete = false;
+    }
+    mEntries[i].mService  = service;
+    mEntries[i].mNodeId   = nodeId;
+    mEntries[i].mFabricId = fabricId;
+    if (deepCopy)
```

Comment:
I guess I screwed up the memory a little bit in the ServicePool. Rewrite with RAII style.

Reply:
I think it's actually OK, because the only place deepCopy is false is during rehashing and that case drops the pool copy without deallocating anything... Might just be worth some comments.